src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -963,14 +962,29 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
         /// <returns>The result of the regex.Replace operation</returns>
         private static object ReplaceOperatorImpl(ExecutionContext context, string input, Regex regex, object substitute)
         {
-            MatchEvaluator matchEvaluator = null;
-            if (!(substitute is string) && LanguagePrimitives.TryConvertTo(substitute, out matchEvaluator))
+            switch (substitute)
             {
-                return regex.Replace(input, matchEvaluator);
-            }
+                case ScriptBlock sb:
+                    MatchEvaluator me = match => {
+                        var result = sb.DoInvokeReturnAsIs(
+                            useLocalScope: false, /* Use local scope to be consistent with 'ForEach/Where-Object {}' and 'collection.ForEach{}/Where{}' */
+                            errorHandlingBehavior: ScriptBlock.ErrorHandlingBehavior.WriteToCurrentErrorPipe,
+                            dollarUnder: match,
+                            input: AutomationNull.Value,
+                            scriptThis: AutomationNull.Value,
+                            args: Utils.EmptyArray<object>());
+
+                        return PSObject.ToStringParser(context, result);;
+                    };
+                    return regex.Replace(input, me);
 
-            string replacement = PSObject.ToStringParser(context, substitute);
-            return regex.Replace(input, replacement);
+                case object val when LanguagePrimitives.TryConvertTo(val, out MatchEvaluator matchEvaluator):
+                    return regex.Replace(input, matchEvaluator);
+
+                default:
+                    string replacement = PSObject.ToStringParser(context, substitute);
```

Comment:
I don't see tests for `PSObject.ToStringParser` - why we use this?

Reply:
It's used in the original code, to convert an object to a string:
```c#
if (rList.Count > 1)
{
    replacement = PSObject.ToStringParser(context, rList[1]);
}
```

Reply:
What PowerShell magic it give us?

Reply:
It converts an object to a string in a powershell way. You can see the code yourself for more information.

Reply:
It's an internal method at https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/MshObject.cs#L1206

Reply:
@daxian-dbw Thanks! I thought that using the method add some magic in the PR but no - it is only to get string.